onnxruntime/core/providers/cpu/tensor/cast_op.cc

```diff
@@ -116,115 +148,117 @@ CastFromString(const std::string& input, DstType& output) {
+    using EigenSrcType = typename EigenType<SrcType>::type;
+    using EigenDstType = typename EigenType<DstType>::type;
+
     const std::ptrdiff_t shape_size = gsl::narrow<std::ptrdiff_t>(shape.Size());
-    const auto in_vector = ConstEigenVectorMap<SrcType>(in.Data<SrcType>(), shape_size);
-    auto out_vector = EigenVectorMap<DstType>(out.MutableData<DstType>(), shape_size);
-    out_vector = in_vector.unaryExpr([](const SrcType& in_scalar) {
-      DstType out_scalar;
-      ScalarCaster<SrcType, DstType>{}.Cast(in_scalar, out_scalar);
-      return out_scalar;
-    });
+    const auto in_vector =
+        ConstEigenVectorMap<EigenSrcType>(reinterpret_cast<const EigenSrcType*>(in.Data<SrcType>()), shape_size);
+    auto out_vector =
+        EigenVectorMap<EigenDstType>(reinterpret_cast<EigenDstType*>(out.MutableData<DstType>()), shape_size);
+    out_vector = in_vector.template cast<EigenDstType>();
   }
 };
 
-template <typename SrcType, typename DstType>
-void CastStringTensor(const Tensor& in, Tensor& out, const TensorShape& shape) {
-  static_assert(std::is_same<SrcType, std::string>::value || std::is_same<DstType, std::string>::value,
-                "Either SrcType or DstType must be std::string.");
-  const std::ptrdiff_t shape_size = gsl::narrow<std::ptrdiff_t>(shape.Size());
-  const auto in_data = in.DataAsSpan<SrcType>();
-  const auto out_data = out.MutableDataAsSpan<DstType>();
-  for (std::ptrdiff_t i = 0; i < shape_size; ++i) {
-    ScalarCaster<SrcType, DstType>{}.Cast(in_data[i], out_data[i]);
-  }
-}
-
 // tensor X -> string
 template <typename SrcType>
 struct TensorCaster<SrcType, std::string> {
-  void Cast(const Tensor& in, Tensor& out, const TensorShape& shape) const {
-    CastStringTensor<SrcType, std::string>(in, out, shape);
+  void Cast(const OpKernelContext&, const Tensor& in, Tensor& out, const TensorShape& shape) const {
+    const std::ptrdiff_t shape_size = gsl::narrow<std::ptrdiff_t>(shape.Size());
+    const auto* in_data = in.Data<SrcType>();
+    auto* out_data = out.MutableData<std::string>();
+    for (std::ptrdiff_t i = 0; i < shape_size; ++i) {
+      CastToString(in_data[i], out_data[i]);
+    }
   }
 };
 
 // tensor string -> X
 template <typename DstType>
 struct TensorCaster<std::string, DstType> {
-  void Cast(const Tensor& in, Tensor& out, const TensorShape& shape) const {
-    CastStringTensor<std::string, DstType>(in, out, shape);
+  void Cast(const OpKernelContext&, const Tensor& in, Tensor& out, const TensorShape& shape) const {
+    const std::ptrdiff_t shape_size = gsl::narrow<std::ptrdiff_t>(shape.Size());
+    const auto* in_data = in.Data<std::string>();
+    auto* out_data = out.MutableData<DstType>();
+    for (std::ptrdiff_t i = 0; i < shape_size; ++i) {
+      CastFromString(in_data[i], out_data[i]);
+    }
   }
 };
 
 #if defined(_M_AMD64)
+// add some specializations to use optimized MLFloat16 -> float conversion
```

Comment:
comment documenting why this is _M_AMD64 only would be good

Reply:
changed to say that this is _M_AMD64 specific. searched the code and from the filenames it looks like it is only implemented for this architecture, but i'm not sure why that is.

---
In reply to: [569866100](https://github.com/microsoft/onnxruntime/pull/6509#discussion_r569866100) [](ancestors = 569866100)

Reply:
Searched the ORT code or Eigen code? Would be good for the comment to mention where you looked and only found amd64 specific optimizations so that in the future we can look in the same places to see if anything has changed and whether other plaforms can have the specializations enabled.

---
In reply to: [569931342](https://github.com/microsoft/onnxruntime/pull/6509#discussion_r569931342) [](ancestors = 569931342,569866100)

Reply:
This is MLAS specific: at one time, the Windows ML team cared about the performance of a FP16 model running on the CPU (for falling back from the DML EP), so I implemented a Windows x64 specific routine to do the conversion. That scenario wasn't important on other platforms, so the kernel was never adapted anywhere else. The scenario may no longer matter either, if you can get support for retiring this from the Windows ML team.

Reply:
ok, thanks for explaining that. the MLAS routine is faster from my measurements, so i think it's good to have this optimized version.

---
In reply to: [570004323](https://github.com/microsoft/onnxruntime/pull/6509#discussion_r570004323) [](ancestors = 570004323)

Reply:
i searched the ORT code for MlasConvertHalfToFloatBuffer.

---
In reply to: [569962495](https://github.com/microsoft/onnxruntime/pull/6509#discussion_r569962495) [](ancestors = 569962495,569931342,569866100)